src/Agent.Plugins/PipelineCache/FingerprintCreator.cs

```diff
@@ -0,0 +1,276 @@
+            }
+            else if (keySegment.First() == '\'')
+            {
+                throw new ArgumentException("A key segment cannot start with a single-quote character`.");
+            }
+            else if (keySegment.First() == '`')
+            {
+                throw new ArgumentException("A key segment cannot start with a backtick character`.");
+            }
+        }
+
+        public static Fingerprint EvaluateKeyToFingerprint(
+            AgentTaskPluginExecutionContext context,
+            string filePathRoot,
+            IEnumerable<string> keySegments)
+        {
+            var sha256 = new SHA256Managed();
+
+            string defaultWorkingDirectory = context.Variables.GetValueOrDefault(
+                "system.defaultworkingdirectory" // Constants.Variables.System.DefaultWorkingDirectory
+                )?.Value;
+
+            var resolvedSegments = new List<string>();
+
+            foreach (string keySegment in keySegments)
+            {
+                CheckKeySegment(keySegment);
+            }
+
+            foreach (string keySegment in keySegments)
+            {
+                if (IsPathy(keySegment))
+                {
+                    context.Verbose($"Interpretting `{keySegment}` as a path.");
+
+                    string[] pathRules = keySegment.Split(new []{','}, StringSplitOptions.RemoveEmptyEntries).Select(s => s.Trim()).ToArray();
+                    string[] includeRules = pathRules.Where(p => !p.StartsWith('!')).ToArray();
+
+                    if (!includeRules.Any())
+                    {
+                        throw new ArgumentException("No include rules specified.");
+                    }
+
+                    var enumerations = new Dictionary<Enumeration,List<string>>();
+                    foreach(string includeRule in includeRules)
+                    {
+                        string workingDirectory = null;
+                        if (!Path.IsPathFullyQualified(includeRule))
+                        {
+                            workingDirectory = defaultWorkingDirectory;
+                        }
+
+                        string absoluteRootRule = MakePathAbsolute(workingDirectory, includeRule);
+                        context.Verbose($"Expanded include rule is `{absoluteRootRule}`.");
+                        Enumeration enumeration = DetermineFileEnumerationFromGlob(absoluteRootRule);
+                        List<string> globs;
+                        if(!enumerations.TryGetValue(enumeration, out globs))
+                        {
+                            enumerations[enumeration] = globs = new List<string>(); 
+                        }
+                        globs.Add(absoluteRootRule);
+                    }
+
+                    string[] excludeRules = pathRules.Where(p => p.StartsWith('!')).ToArray();
+                    string[] absoluteExcludeRules = excludeRules.Select(excludeRule => {
```

Comment:
I'd say we leverage off a public dotnet globbing library:
[https://github.com/dazinator/DotNet.Glob](https://github.com/dazinator/DotNet.Glob)

Reply:
or do we match the rest of DevOps? 🤔 @willsmythe?

---
In reply to: [306972347](https://github.com/microsoft/azure-pipelines-agent/pull/2358#discussion_r306972347) [](ancestors = 306972347)